tests/qa_automation/qa_run.pm

```diff
@@ -31,71 +33,69 @@ sub test_suite() {
     die "you need to overload test_suite in your class";
 }
 
+sub wait_testrun {
+    my ($interval) = @_;
+    $interval = 300 unless defined $interval;
+
+    my $done_file = '/var/log/qaset/control/DONE';
+    while (1) {
+        my $pattern = "TESTRUN_FINISHED-" . int(rand(999999));
+        type_string "sleep 0.1 ; [[ -f $done_file ]] && echo '$pattern' | tee /dev/$serialdev\n";
+        my $ret = wait_serial($pattern, $interval);
+        if ($ret) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+sub qa_upload_logs {
+    my ($dir, $pattern) = @_;
+    my $output = script_output("find '$dir' -type f -name '$pattern'", 300);
+    my @log_files = split("\n", $output);
+    # upload logs
+    foreach my $log_file (@log_files) {
+        $log_file =~ s/^\s+|\s+$//g;
+        upload_logs($log_file);
+    }
+}
+
 # qa_testset_automation validation test
 sub run() {
     my $self = shift;
-
+    # system boot & login
     assert_screen "inst-bootmenu", 30;
-    send_key "ret";    # boot
-
+    send_key "ret";
     assert_screen "grub2", 15;
     send_key "ret";
-
     assert_screen "text-login", 50;
     type_string "root\n";
     assert_screen "password-prompt", 10;
     type_password;
     type_string "\n";
-    sleep 1;
-
-    # Add Repo - http://dist.nue.suse.com/ibs/QA:/Head/SLE-12-SP1/
+    sleep 2;
+    # Repo & Package
     assert_script_run "zypper --no-gpg-check -n ar -f " . get_var('QA_HEAD_REPO') . " qa_ibs";
```

Comment:
Add step to print all repository is helpful for debugging when pacakge installation failed.